firstpartydata/first_party_data_test.go

```diff
@@ -0,0 +1,836 @@
+				"site": {},
+			},
+		},
+		{
+			description: "No FPD present in req",
+			input: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  }
+  				},
+  				"app": {
+  				  "id": "appId",
+  				  "ext": {
+                  }
+  				},
+			  	"user": {
+  				  "id": "userId",
+				  "ext": {}
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			output: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  }
+  				},
+  				"app": {
+  				  "id": "appId",
+				  "ext": {}
+  				},
+				"user": {
+  				  "id": "userId",
+				  "ext": {}
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			expectedFpd: map[string][]byte{
+				"app":  {},
+				"user": {},
+				"site": {},
+			},
+		},
+		{
+			description: "Site FPD only present",
+			input: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  },
+				  "ext": {
+					"data": {"someappfpd": true},
+					"amp": 1
+                  }
+  				},
+  				"app": {
+  				  "id": "appId"
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			output: []byte(`{
+  				"id": "bid_id",
+  				"site": {
+  				  "id":"reqSiteId",
+  				  "page": "http://www.foobar.com/1234.html",
+  				  "publisher": {
+  				    "id": "1"
+  				  },
+                 "ext": {
+					"amp": 1
+                  }
+  				},
+  				"app": {
+  				  "id": "appId"
+  				},
+  				"tmax": 5000,
+  				"source": {
+  				  "tid": "ad839de0-5ae6-40bb-92b2-af8bad6439b3"
+  				}
+			}`),
+			expectedFpd: map[string][]byte{
+				"app":  {},
+				"user": {},
+				"site": []byte(`{"someappfpd": true}`),
+			},
+		},
+	}
+	for _, test := range testCases {
+		var inputTestReq openrtb_ext.RequestWrapper
+		err := json.Unmarshal(test.input, &inputTestReq)
+		assert.NoError(t, err, "Error should be nil")
+
+		fpd, err := ExtractGlobalFPD(&inputTestReq)
+		inputTestReq.RebuildRequest()
+		assert.NoError(t, err, "Error should be nil")
+
+		var outputTestReq openrtb_ext.RequestWrapper
+		err = json.Unmarshal(test.output, &outputTestReq)
+		assert.NoError(t, err, "Error should be nil")
+
+		if fpd[userKey] != nil {
+			if string(inputTestReq.User.Ext) != "" && string(outputTestReq.User.Ext) != "" {
+				assert.JSONEq(t, string(inputTestReq.User.Ext), string(outputTestReq.User.Ext), "Result is incorrect")
+			}
+			assert.Equal(t, test.expectedFpd[userKey], fpd[userKey], "FPD is incorrect")
+		}
+		if fpd[appKey] != nil {
```

Comment:
Similar comments as above for app and site related assertions here

Reply:
Fixed